### PR TITLE
Month::overflowing_add/sub

### PR DESCRIPTION
**Pull request by esheppa**
Add an overflowing_add/sub similar to the equivalents for `NaiveTime`, which are useful for `NaiveDate` + `Months`
Also change `Months` constructor to only accept `u32 <= i32::MAX` and add a `try_new`

### CODE REVIEW

**Comment by esheppa** [comment]
If this is looking like a good API, I'll add some more documentation prior to this being merged

**Comment by esheppa** [comment]
I still need to do the `MIN_YEAR` / `MAX_YEAR` checks in `diff_months` as well

File: src/naive/date.rs [hunk 1]
```diff
@@ -643,53 +640,28 @@ impl NaiveDate {
...
+        let month: crate::Month = num_traits::FromPrimitive::from_u32(self.month()).unwrap();
 
-            (year + 1, month - 12)
-        } else {
-            (year, month)
+        let (years, month) = match forwards {
+            true => month.overflowing_add(months),
+            false => month.overflowing_sub(months),
         };
+        let year = self.year() + years;
```
Comment (djc) [inline src/naive/date.rs:655]: No more overflow/underflow checking for the years?

File: src/naive/date.rs [hunk 2]
```diff
@@ -643,53 +640,28 @@ impl NaiveDate {
...
-
-            (year - 1, month + 12)
-        } else if month > 12 {
-            if year == MAX_YEAR {
-                return None;
-            }
+    fn diff_months(self, months: Months, forwards: bool) -> Option<Self> {
+        // safe as .month() returns [1..12]
+        // can be further improved when .month() returns a Month
+        let month: crate::Month = num_traits::FromPrimitive::from_u32(self.month()).unwrap();
```
Comment (djc) [inline src/naive/date.rs:649]: I don't see the point of using `num_traits::FromPrimitive` and using `unwrap()` if we can just use `as` here.

File: src/month.rs [hunk 3]
```diff
@@ -195,8 +223,23 @@ pub struct Months(pub(crate) u32);
 
 impl Months {
     /// Construct a new `Months` from a number of months
+    pub fn new_opt(num: u32) -> Option<Self> {
+        if num <= core::i32::MAX as u32 {
```
Comment (djc) [inline src/month.rs:227]: I'm not really a fan of this as the constraint ends up not being expressed in the type system, so we have to lean on convention/remembering/understanding. IMO it's better just to it as `u32` (which also makes the initialization API simple) and check the range in usage context.

**Review by djc — COMMENTED** [review]
In principle I'm open to the modularization of moving some of `diff_months()`'s internals into public API on `Month`, but I'm not sure this API is really natural/otherwise useful.

File: src/naive/date.rs [hunk 4]
```diff
@@ -2153,26 +2125,13 @@ mod tests {
     use std::{i32, u32};
 
     #[test]
-    fn diff_months() {
+    fn test_diff_months() {
```
Comment (djc) [inline src/naive/date.rs:2128]: What does the `test_` prefix buy us? This seems like an unrelated change.

File: src/naive/date.rs [hunk 5]
```diff
@@ -643,53 +640,28 @@ impl NaiveDate {
...
-            (year, month)
+        let (years, month) = match forwards {
+            true => month.overflowing_add(months),
+            false => month.overflowing_sub(months),
         };
+        let year = self.year() + years;
+        let month = month.number_from_month();
 
         // Clamp original day in case new month is shorter
-
```
Comment (djc) [inline src/naive/date.rs:686]: I'd prefer to keep the empty line here

File: src/naive/date.rs (hunk 2 above)
Comment (esheppa) [reply src/naive/date.rs:649]: I need the actual `Month` type here as that's what `overflowing_add/sub` take as the `&self` parameter. This will be able to be cleaned up later if/when we fix #727

File: src/naive/date.rs (hunk 1 above)
Comment (esheppa) [reply src/naive/date.rs:655]: Accidentally removed too much from the function. I've added it back as:

```rust
        if year < MIN_YEAR || year > MAX_YEAR {
            return None;
        }
```

Or do you mean the `self.year() + years` overflowing/underflowing the `i32`?

File: src/month.rs (hunk 3 above)
Comment (esheppa) [reply src/month.rs:227]: My thinking with this is that if the `Months` is invalid, its better for this to show up when creating the months rather than adding it, but admittedly most `Months` will be made and added as part of the same expression. Perhaps another option is taking `i32` here but returning `None` on negative values? Or else removing `new_opt` and having the check be done inside `overflowing_add/sub`. This also forces months to have one less possible value on the sub side due to `i32::MAX != abs(i32::MIN)`, but potentially this is acceptable as it is erring on the side of allowing a slightly smaller maximum subtraction?

File: src/naive/date.rs (hunk 1 above)
Comment (djc) [reply src/naive/date.rs:655]: Both...

File: src/month.rs (hunk 3 above)
Comment (djc) [reply src/month.rs:227]: But that's just the point: a `Months::new(u32::MAX)` is not invalid by itself -- it's conceptually perfectly valid. It's only invalid in the context of what `checked_add_months()`/`checked_sub_months()` are trying to do, as such IMO the burden of validation should fall on the context, not the type itself.

**Comment by esheppa** [comment]
With regard to the API, I agree that it is debatable if it is generally useful. A couple of potential use cases are:
* Users wanting to implement `add_months` but with their own clamping behavior on invalid days
* Users wanting to know the month n months after another, irrespective of the year

Alternatively if you like this but don't think think the public API is useful, it could be made private?

File: src/naive/date.rs (hunk 1 above)
Comment (esheppa) [reply src/naive/date.rs:658]: will add the check on `i32` as well

File: src/naive/date.rs (hunk 2 above)
Comment (esheppa) [reply src/naive/date.rs:649]: I don't think it is possible to go from `u32` to an enumeration thought?

File: src/month.rs (hunk 3 above)
Comment (esheppa) [reply src/month.rs:245]: Makes sense, will move the checks there

**Comment by djc** [comment]
I think this is only good if the public API is useful. Otherwise the indirection makes the code harder to understand and I'd prefer to keep it as I had it in my original PR. It sounds to me like we don't yet have a very convincing use case in mind for exposing this publicly, so in that case I think we should leave this for now and keep this in mind as a potential direction for future evolution.

**Comment by djc** [comment]
(Sorry, closed it accidentally.)
